compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -207,8 +208,19 @@ impl<Prov> Allocation<Prov> {
         align: Align,
         mutability: Mutability,
     ) -> Self {
-        let bytes = Box::<[u8]>::from(slice.into());
-        let size = Size::from_bytes(bytes.len());
+        let slice: Cow<'a, [u8]> = slice.into();
+        let size = Size::from_bytes(slice.len());
+        let align_usize: usize = align.bytes().try_into().unwrap();
+        let layout = std::alloc::Layout::from_size_align(slice.len(), align_usize).unwrap();
+        let bytes = unsafe {
```

Comment:
I think my biggest concern is that this introduces unsafe code deep in the core of the interpreter engine. This is a rather significant cost and risk that the interpreter has to carry, even the compile-time interpreter that definitely does not want or need this.

At the least, we should find a way to factor this such that all the unsafe code lives in the Miri repository. Even that is still not great though, it is exactly the kind of global cost I was worried about.

Reply:
Ah, I see your point, but I don't think we can align the bytes of the allocation without `unsafe` code. 
If we want to factor this out into Miri, then one option could be to refactor `Allocation` to expose `bytes` so that the `bytes` alignment can be done on the Miri side. Although, this would be a bigger change to `Allocation`. 
What do you think?

Reply:
No we certainly don't want to expose `bytes`. But we can make `Allocation` generic over how bytes are allocated.

You can add a new type parameter for that:
```rust
pub struct Allocation<Prov = AllocId, Extra = (), Bytes = Box<[u8]>> {
  bytes: Bytes
```
and then define a suitable trait
```rust
trait AllocationBytes
```
and add a bound `Bytes: AllocationBytes` whereveer needed.

The trait needs to then contain all the operations you need. The implementation in rustc can use `Box<[u8]>` and be completely safe, and that way only Miri needs unsafe code.

Reply:
Ah that's smart -- I had tried earlier to make `Allocation` generic over the allocator for bytes which didn't work. But this makes sense!